README
README
February 10, 2017

==============================================================================
Most end users would probably only need to update the /etc/X11/xorg.conf file 
and sometime probably the Wacom X driver as well to use the rich features that 
Wacom tablets offer.  The steps could be as simple as:

    $ su
    # cp /etc/X11/xorg.conf /etc/X11/xorg.conf.bak
    # gedit /etc/X11/xorg.conf

then follow the step listed at http://linuxwacom.sourceforge.net/index_old.php/howto/x11
or "man wacom" if you don't have network access to add sections for Wacom device.

However, if you have a USB tablet which is newer than your running kernel, i.e., 
the tablet is not supported by your running kernel, or you are a developer who is 
willing to build the driver from the source, there are extra work waiting for you.

==============================================================================

PREREQUISITES:

In order to build the linuxwacom X and kernel drivers, several GNU utilities will
need to be present on your system. Distributions like Fedora provide a handy
"Development Tools" group that will bring in all the necessary software. In other
distributions, you may need to install the necessary packages individually. For
users of Solaris it is important to use the GNU versions of these utilities, e.g.
as installed from OpenCSW; the SUNW versions of these utilities are incompatible.

  * GCC4
  * GNU make
  * GNU sed
  * pkgconfig
  * GNU libtool  (Only required for builds from git)
  * GNU autoconf (Only required for builds from git)
  * GNU automake (Only required for builds from git)

Additionally, a number of development headers will need to be made available for
the drivers to compile. Some distributions (e.g. RHEL 5) provide these headers as
installable packages, while others (e.g. Solaris) may require you to locate
appropriate headers elsewhere.

  * RHEL 5:
      * xorg-x11-server-sdk
          * libXau-devel
          * xorg-x11-util-macros
          * xorg-x11-proto-devel
          * libXdmcp-devel
          * mesa-libGL-devel
          * libX11-devel
      * libXi-devel
      * kernel-devel

  * Other:
      * xorg-server [e.g. https://www.x.org/releases/X11R7.2/src/xserver/xorg-server-X11R7.2-1.2.0.tar.bz2]
      * xproto      [e.g. https://www.x.org/releases/X11R7.2/src/proto/xproto-X11R7.2-7.0.10.tar.bz2]
      * libX11      [e.g. https://www.x.org/releases/X11R7.2/src/lib/libX11-X11R7.2-1.1.1.tar.bz2]
      * libXi       [e.g. https://www.x.org/releases/X11R7.2/src/lib/libXi-X11R7.2-1.0.2.tar.bz2]

Other optional dependencies may also need to be installed in order to build some
pieces of software.

  * tcl  (Required for ????)
  * tk   (Required for ????)


BUILDING FROM SOURCE:

As you would expect, run configure and make. If building from git, it is necessary
to first run `./bootstrap` to generate the configure script. Configure options are
explained below.

    $ ./configure && make


CONFIGURE OPTIONS:

This file contains hints for building the sources.  There are a number of build
 targets, some of which are not built automatically.  Here is the
breakdown:

      wacom.o: not built by default (--enable-wacom, --disable-wacom)
      wacdump: built by default (--enable-wacdump, --disable-wacdump)
       xidump: built by default (--enable-xidump, --disable-xidump)
    xsetwacom: built by default (--enable-xsetwacom, --disable-xsetwacom)
        hid.o: not built by default (--enable-hid, --disable-hid;  Only kernels 2.6.18 or older need it.)
  wacom_drv.o: built by default (--enable-wacomdrv, --disable-wacomdrv)

There are many options, and configure will attempt to guess the correct
values for most of them.  By default, only wacomcpl, xsetwacom,  wacom_drv.o, 
xidump, and wacdump program are built.

Please surf the HOWTO page at http://linuxwacom.sourceforge.net/index_old.php/howto/main
(or the miniHOWTO at http://linuxwacom.sourceforge.net/index_old.php/minihowto if 
you run a Fedora system) to find the proper options if the default building 
environment doesn't work for you.

==============================================================================
Users who intend to build the driver for Solaris should be aware that this is a
minimally-tested configuration. The ability to build under Solaris may break
without warning. Additionally, it is highly recommended to use prerequisite
packages from the OpenCSW archive as these are also used by the developers.

==============================================================================

WEBSITE: http://linuxwacom.sf.net


ChangeLog
2012-01-14  Ping Cheng <pingc@wacom.com>
	* Add Cinitq 24HD support
	* Label 0.9.3

2011-10-12  Ping Cheng <pingc@wacom.com>
	* Add LED support for Intuos4 and Cinitq 21UX2
	* Label 0.9.1

2011-06-20  Ping Cheng <pingc@wacom.com>
	* Add "-lc" to Makefile.am (Daniel Grafe)
	* Add "ignore first USB motion events" back. Otherwise cursor jumps when first in prox
	* Updated wacdump.php with input from Patrick Horgan and David Foley
	* Updated "Building wacom driver On RHEL 5" page by Sylvain Giasson
	* Label 0.9.0

2011-04-27  Ping Cheng <pingc@wacom.com>
	* Put wacom_ioctl back for OLED (Eduard Hasenleithner)
	* Fixed error path in wacom_probe (Alexander Strakh)
	* Updated wacomcfg.c for touch device (David Foley)
	* Some minor kernel updated
	* Label 0.8.9

2011-03-03  Ping Cheng <pingc@wacom.com>
	* Always retrieve (x,y) from the kernel the first time in prox for non-pad tool
	* Don't read data while driver loading/unloading to avoid a hotplugging race condition
	* Don't turn device off when getting read errors
	* Fixed a missing tool type for Cintiq in 2.6.16/wacom_wac.c
	* Label 0.8.8-11

2010-10-12  Ping Cheng <pingc@wacom.com>
	* Fixed a kernel bug for Cintiq 21UX2 pressure (Aristeu Rozanski <aris@redhat.com>)
	* Fixed a print format error in xidump.c (Aristeu Rozanski)
	* Label 0.8.8-10

2010-10-01  Ping Cheng <pingc@wacom.com>
	* Specify the tools supported by Intuos and CintiqV5 
	  series respectively in the kernel
	* Calibration-only wacomcpl has the options to display warnings or 
	  flashing the crosshairs in red when abnormal (x,y) is detected
	* Enable all tools for X server 1.4 and older and kernel 2.6.24 and older
	* Peter Hutterer's configure message update for newer Xorg
	* Label 0.8.8-9

2010-07-27  Ping Cheng <pingc@wacom.com>
	* Updated kernel driver for 2.6.23 and later
	* Label 0.8.8-8

2010-07-22  Ping Cheng <pingc@wacom.com>
	* Fixed a memory leakage in util/wacomcfg.c (Eldar Firing <eldarf@sf.net>)
	* Replaced usb_buffer_free with usb_free_coherent for 2.6.35 
		- by Diego Calleja <diegocg@gmail.com>
	* Removed a few unused routines in xf86Wacom.c
	* Label 0.8.8-7

2010-07-14  Ping Cheng <pingc@wacom.com>
	* Updated udev wacom rules
	* Updated web pages for new devices
	* Added call EVIOCGKEY back in wcmUSB.c
	* Label 0.8.8-6

2010-07-12  Ping Cheng <pingc@wacom.com>
	* Updated configure.in and src/Makefile.am to include xorg server 1.6.5
	* Updated wcmUSB.c to use default nbuttons and npadkeys since kernels 
	  older than 2.6.24 doesn't return proper keys from EVIOCGKEY
	* Added a new xsetwacom option, isDisplay, to report if the tablet is 
	  a display tablet or not 
	* Label 0.8.8-5

2010-06-29  Ping Cheng <pingc@wacom.com>
	* Updated configure.in and src/Makefile.am for XFree86 servers
	* Updated 2.6.16/wacom_wac.c for kernels older than 2.6.21
	* Added "\" to wacomcpl-exec to continue the partial lines
	* Label 0.8.8-4

2010-06-18  Ping Cheng <pingc@wacom.com>
	* Updated configure.in for older X servers
	* Updated 2.6.16/wacom_wac.c for Cintiq 21UX2
	* Label 0.8.8-3

2010-06-16  Ping Cheng <pingc@wacom.com>
	* Updated wacom_wac.c and wacom_wac.h for DTU-2231 and DTU-1631
	* Updated wacomcpl
	* Label 0.8.8-2

2010-06-11  Ping Cheng <pingc@wacom.com>
	* Fixed a few minor issues
	* Added DTU-2231 support
	* Added DTU-1631 support
	* Label 0.8.8-1

2010-06-07  Ping Cheng <pingc@wacom.com>
	* Fixed free toolarea issue
	* Moved "retrieving tools" from wcmCommon to wcmUSB, where it belongs
	* Initialize common->npadkeys before recounting the keys
	* Label 0.8.8

2010-06-02  Ping Cheng <pingc@wacom.com>
	* Added the xsetwacom manpage
		- by Peter Hutterer <peter.hutterer@redhat.com>
	* Free common, common->toollist, priv->tool, and toolarea
	* Added Application Defined functionality for buttons
	* Retrieve supported tools from the kernel if a tool is on 
	  the tablet when X server started
	* Label 0.8.7-2

2010-05-21  Ping Cheng <pingc@wacom.com>
	* Normalized the 6th valuator to 2048
	  this valuator reports absolute wheel or ArtPen rotation
	  they are normally considered as the tangential pressure
	* Updated wacomcpl for button options
	* Added support for Intuos4WL (USB ndpoint)E in the kernel
	* Added support for Intuos4WL in xdrv (both bluetooth and USB)
	* Label 0.8.7-1

2010-05-17  Ping Cheng <pingc@wacom.com>
	* Restructured supported kernesl
	* Removed kernels 2.6.9 to 2.6.15 support
	* Added support for Cintiq 21UX2
	* Label 0.8.7

2010-05-12  Ping Cheng <pingc@wacom.com>
	* Updated configure.in to disable wacom_drv and its utilities
	  and tell user to build wacom_drv from xf86-input-wacom for 
	  Xorg server 1.7 or later
	* Label 0.8.6-2
	  
2010-04-30  Ping Cheng <pingc@wacom.com>
	* Updated wacom_sys.c for system resume
	* Reset wcmTouchpadMode in xf86WcmSoftOutEvent
	* Added wcmCustomDebug.c for extra debug information
		- by Stefan Schimanski
	* Label 0.8.6-1

2010-04-09  Ping Cheng <pingc@wacom.com>
	* Updated calibration in wacomcpl (Stefan Schimanski)
	* Updated wacom_wac.c for 2.6.24+ 
		- by Dmitry Torokhov <dmitry.torokhov@gmail.com>
	* Updated pressure normalization
	* Updated ISDV4 data parsing routine
	* Updated wacom man page
	* Updated linuxwacom web pages
	* Label 0.8.6

2010-03-26  Ping Cheng <pingc@wacom.com>
	* Updated kernel driver for 2FGT and Bamboo touch
	* Added 2.6.30 path for kernels 2.6.30 and later
	* Normalized pressure sensitivity to 2048
	* Updated 2FGT gesture with Chris Bagwell's suggestion
	* Label 0.8.5-12

2010-03-10  Ping Cheng <pingc@wacom.com>
	* Add modifiers (right) to wcmAction.c
	* Fixed a copy/paste error in the license statement
	* Support kernel 2.6.33
	* Cleaned kernel drivers for 2FGT
	* Add HAL_CFLAGS to CFLAGS 
		- by Pedro Gimeno <pg-lwd391@personal.formauri.es>
	* Retrieve tablet_id early on in wcmValidateDevice
	* Call getRange for both touch and pad, whichever comes the first
		- by Chris Bagwell <chris@cnpbagwell.com>
	* Fixed a keystroke bug for non-core mode of older Xorgs
		- by Stefan Schimanski

src/mousedev.c
/*
 * $Id: mousedev.c,v 1.3 2005-06-16 17:04:09 pingc Exp $
 *
 *  Copyright (c) 1999-2000 Vojtech Pavlik
 *  Copyright (c) 2003 John Joganic         <john@joganic.com>
 *
 *  Input driver to ImExPS/2 device driver module.
 *
 *  Sponsored by SuSE
 *
 *  Revision History - AS BRANCH FROM LINUX WACOM PROJECT
 *  2003-03-31  2.4.20-j0.5.0  - Added to Linux Wacom project
 *                               Applied exception for wacom tablets
 * 
 */

/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 * 
 * Should you need to contact me, the author, you can do so either by
 * e-mail - mail your message to <vojtech@suse.cz>, or by paper mail:
 * Vojtech Pavlik, Ucitelska 1576, Prague 8, 182 00 Czech Republic
 */

#define MOUSEDEV_VERSION "2.4.20-j0.5.0"

#include <linux/autoconf.h>
#if defined(CONFIG_MODVERSIONS) && !defined(MODVERSIONS)
#   define MODVERSIONS
#endif

#ifdef MODVERSIONS
#include <linux/modversions.h>
#endif

#define MOUSEDEV_MINOR_BASE 	32
#define MOUSEDEV_MINORS		32
#define MOUSEDEV_MIX		31

#include <linux/slab.h>
#include <linux/poll.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/input.h>
#include <linux/config.h>
#include <linux/smp_lock.h>
#include <linux/random.h>

#ifndef CONFIG_INPUT_MOUSEDEV_SCREEN_X
#define CONFIG_INPUT_MOUSEDEV_SCREEN_X	1024
#endif
#ifndef CONFIG_INPUT_MOUSEDEV_SCREEN_Y
#define CONFIG_INPUT_MOUSEDEV_SCREEN_Y	768
#endif

struct mousedev {
	int exist;
	int open;
	int minor;
	wait_queue_head_t wait;
	struct mousedev_list *list;
	struct input_handle handle;
	devfs_handle_t devfs;
};

struct mousedev_list {
	struct fasync_struct *fasync;
	struct mousedev *mousedev;
	struct mousedev_list *next;
	int dx, dy, dz, oldx, oldy;
	signed char ps2[6];
	unsigned long buttons;
	unsigned char ready, buffer, bufsiz;
	unsigned char mode, imexseq, impsseq;
};

#define MOUSEDEV_SEQ_LEN	6

static unsigned char mousedev_imps_seq[] = { 0xf3, 200, 0xf3, 100, 0xf3, 80 };
static unsigned char mousedev_imex_seq[] = { 0xf3, 200, 0xf3, 200, 0xf3, 80 };

static struct input_handler mousedev_handler;

static struct mousedev *mousedev_table[MOUSEDEV_MINORS];
static struct mousedev mousedev_mix;

static int xres = CONFIG_INPUT_MOUSEDEV_SCREEN_X;
static int yres = CONFIG_INPUT_MOUSEDEV_SCREEN_Y;

static void mousedev_event(struct input_handle *handle, unsigned int type, unsigned int code, int value)
{
	struct mousedev *mousedevs[3] = { handle->private, &mousedev_mix, NULL };
	struct mousedev **mousedev = mousedevs;
	struct mousedev_list *list;
	int index, size;

	add_mouse_randomness((type << 4) ^ code ^ (code >> 4) ^ value);

	while (*mousedev) {
		list = (*mousedev)->list;
		while (list) {
			switch (type) {
				case EV_ABS:
					if (test_bit(BTN_TRIGGER, handle->dev->keybit))
						break;
					switch (code) {
						case ABS_X:	
							size = handle->dev->absmax[ABS_X] - handle->dev->absmin[ABS_X];
							list->dx += (value * xres - list->oldx) / size;
							list->oldx += list->dx * size;
							break;
						case ABS_Y:
							size = handle->dev->absmax[ABS_Y] - handle->dev->absmin[ABS_Y];
							list->dy -= (value * yres - list->oldy) / size;
							list->oldy -= list->dy * size;
							break;
					}
					break;

				case EV_REL:
					switch (code) {
						case REL_X:	list->dx += value; break;
						case REL_Y:	list->dy -= value; break;
						case REL_WHEEL:	if (list->mode) list->dz -= value; break;
					}
					break;

				case EV_KEY:
					switch (code) {
						case BTN_0:
						case BTN_TOUCH:
						case BTN_LEFT:   index = 0; break;
						case BTN_4:
						case BTN_EXTRA:  if (list->mode == 2) { index = 4; break; }
						case BTN_STYLUS:
						case BTN_1:
						case BTN_RIGHT:  index = 1; break;
						case BTN_3:
						case BTN_SIDE:   if (list->mode == 2) { index = 3; break; }
						case BTN_2:
						case BTN_STYLUS2:
						case BTN_MIDDLE: index = 2; break;	
						default: return;
					}
					switch (value) {
						case 0: clear_bit(index, &list->buttons); break;
						case 1: set_bit(index, &list->buttons); break;
						case 2: return;
					}
					break;
			}
					
			list->ready = 1;

			kill_fasync(&list->fasync, SIGIO, POLL_IN);

			list = list->next;
		}

		wake_up_interruptible(&((*mousedev)->wait));
		mousedev++;
	}
}

static int mousedev_fasync(int fd, struct file *file, int on)
{
	int retval;
	struct mousedev_list *list = file->private_data;
	retval = fasync_helper(fd, file, on, &list->fasync);
	return retval < 0 ? retval : 0;
}

static int mousedev_release(struct inode * inode, struct file * file)
{
	struct mousedev_list *list = file->private_data;
	struct mousedev_list **listptr;

	lock_kernel();
	listptr = &list->mousedev->list;
	mousedev_fasync(-1, file, 0);

	while (*listptr && (*listptr != list))
		listptr = &((*listptr)->next);
	*listptr = (*listptr)->next;

	if (!--list->mousedev->open) {
		if (list->mousedev->minor == MOUSEDEV_MIX) {
			struct input_handle *handle = mousedev_handler.handle;
			while (handle) {
				struct mousedev *mousedev = handle->private;
				if (!mousedev->open) {
					if (mousedev->exist) {
						input_close_device(&mousedev->handle);
					} else {
						input_unregister_minor(mousedev->devfs);
						mousedev_table[mousedev->minor] = NULL;
						kfree(mousedev);
					}
				}
				handle = handle->hnext;
			}
		} else {
			if (!mousedev_mix.open) {
				if (list->mousedev->exist) {
					input_close_device(&list->mousedev->handle);
				} else {
					input_unregister_minor(list->mousedev->devfs);
					mousedev_table[list->mousedev->minor] = NULL;
					kfree(list->mousedev);
				}
			}
		}
	}
	
	kfree(list);
	unlock_kernel();

	return 0;
}

static int mousedev_open(struct inode * inode, struct file * file)
{
	struct mousedev_list *list;
	int i = MINOR(inode->i_rdev) - MOUSEDEV_MINOR_BASE;

	if (i >= MOUSEDEV_MINORS || !mousedev_table[i])
		return -ENODEV;

	if (!(list = kmalloc(sizeof(struct mousedev_list), GFP_KERNEL)))
		return -ENOMEM;
	memset(list, 0, sizeof(struct mousedev_list));

	list->mousedev = mousedev_table[i];
	list->next = mousedev_table[i]->list;
	mousedev_table[i]->list = list;
	file->private_data = list;

	if (!list->mousedev->open++) {
		if (list->mousedev->minor == MOUSEDEV_MIX) {
			struct input_handle *handle = mousedev_handler.handle;
			while (handle) {
				struct mousedev *mousedev = handle->private;
				if (!mousedev->open)
					if (mousedev->exist)	
						input_open_device(handle);
				handle = handle->hnext;
			}
		} else {
			if (!mousedev_mix.open)
				if (list->mousedev->exist)	
					input_open_device(&list->mousedev->handle);
		}
	}

	return 0;
}

static void mousedev_packet(struct mousedev_list *list, unsigned char off)
{
	list->ps2[off] = 0x08 | ((list->dx < 0) << 4) | ((list->dy < 0) << 5) | (list->buttons & 0x07);
	list->ps2[off + 1] = (list->dx > 127 ? 127 : (list->dx < -127 ? -127 : list->dx));
	list->ps2[off + 2] = (list->dy > 127 ? 127 : (list->dy < -127 ? -127 : list->dy));
	list->dx -= list->ps2[off + 1];
	list->dy -= list->ps2[off + 2];
	list->bufsiz = off + 3;

	if (list->mode == 2) {
		list->ps2[off + 3] = (list->dz > 7 ? 7 : (list->dz < -7 ? -7 : list->dz));
		list->dz -= list->ps2[off + 3];
		list->ps2[off + 3] = (list->ps2[off + 3] & 0x0f) | ((list->buttons & 0x18) << 1);
		list->bufsiz++;
	}
	
	if (list->mode == 1) {
		list->ps2[off + 3] = (list->dz > 127 ? 127 : (list->dz < -127 ? -127 : list->dz));
		list->dz -= list->ps2[off + 3];
		list->bufsiz++;
	}

	if (!list->dx && !list->dy && (!list->mode || !list->dz)) list->ready = 0;
	list->buffer = list->bufsiz;
}


static ssize_t mousedev_write(struct file * file, const char * buffer, size_t count, loff_t *ppos)
{
	struct mousedev_list *list = file->private_data;
	unsigned char c;
	int i;

	for (i = 0; i < count; i++) {

		if (get_user(c, buffer + i))
			return -EFAULT;

		if (c == mousedev_imex_seq[list->imexseq]) {
			if (++list->imexseq == MOUSEDEV_SEQ_LEN) {
				list->imexseq = 0;
				list->mode = 2;
			}
		} else list->imexseq = 0;

		if (c == mousedev_imps_seq[list->impsseq]) {
			if (++list->impsseq == MOUSEDEV_SEQ_LEN) {
				list->impsseq = 0;
				list->mode = 1;
			}
		} else list->impsseq = 0;

		list->ps2[0] = 0xfa;
		list->bufsiz = 1;
		list->ready = 1;

		switch (c) {

			case 0xeb: /* Poll */
				mousedev_packet(list, 1);
				break;

			case 0xf2: /* Get ID */
				switch (list->mode) {
					case 0: list->ps2[1] = 0; break;
					case 1: list->ps2[1] = 3; break;
					case 2: list->ps2[1] = 4; break;
				}
				list->bufsiz = 2;
				break;

			case 0xe9: /* Get info */
				list->ps2[1] = 0x60; list->ps2[2] = 3; list->ps2[3] = 200;
				list->bufsiz = 4;
				break;
		}

		list->buffer = list->bufsiz;
	}

	kill_fasync(&list->fasync, SIGIO, POLL_IN);

	wake_up_interruptible(&list->mousedev->wait);
		
	return count;
}

static ssize_t mousedev_read(struct file * file, char * buffer, size_t count, loff_t *ppos)
{
	DECLARE_WAITQUEUE(wait, current);
	struct mousedev_list *list = file->private_data;
	int retval = 0;

	if (!list->ready && !list->buffer) {

		add_wait_queue(&list->mousedev->wait, &wait);
		current->state = TASK_INTERRUPTIBLE;

		while (!list->ready) {

			if (file->f_flags & O_NONBLOCK) {
				retval = -EAGAIN;
				break;
			}
			if (signal_pending(current)) {
				retval = -ERESTARTSYS;
				break;
			}

			schedule();
		}

		current->state = TASK_RUNNING;
		remove_wait_queue(&list->mousedev->wait, &wait);
	}

	if (retval)
		return retval;

	if (!list->buffer)
		mousedev_packet(list, 0);

	if (count > list->buffer)
		count = list->buffer;

	if (copy_to_user(buffer, list->ps2 + list->bufsiz - list->buffer, count))
		return -EFAULT;
	
	list->buffer -= count;

	return count;	
}

/* No kernel lock - fine */
static unsigned int mousedev_poll(struct file *file, poll_table *wait)
{
	struct mousedev_list *list = file->private_data;
	poll_wait(file, &list->mousedev->wait, wait);
	if (list->ready || list->buffer)
		return POLLIN | POLLRDNORM;
	return 0;
}

struct file_operations mousedev_fops = {
	owner:		THIS_MODULE,
	read:		mousedev_read,
	write:		mousedev_write,
	poll:		mousedev_poll,
	open:		mousedev_open,
	release:	mousedev_release,
	fasync:		mousedev_fasync,
};

static struct input_handle *mousedev_connect(struct input_handler *handler, struct input_dev *dev)
{
	struct mousedev *mousedev;
	int minor = 0;

	/* Ignore all wacom tablets */
	if (dev->idvendor == 0x56a)
	{
		printk(KERN_INFO "mousedev.c: Ignoring wacom tablet\n");
		return NULL;
	}

	if (!test_bit(EV_KEY, dev->evbit) ||
	   (!test_bit(BTN_LEFT, dev->keybit) && !test_bit(BTN_TOUCH, dev->keybit)))
		return NULL;

	if ((!test_bit(EV_REL, dev->evbit) || !test_bit(REL_X, dev->relbit)) &&
	    (!test_bit(EV_ABS, dev->evbit) || !test_bit(ABS_X, dev->absbit)))
		return NULL;

	for (minor = 0; minor < MOUSEDEV_MINORS && mousedev_table[minor]; minor++);
	if (minor == MOUSEDEV_MINORS) {
		printk(KERN_ERR "mousedev: no more free mousedev devices\n");
		return NULL;
	}

	if (!(mousedev = kmalloc(sizeof(struct mousedev), GFP_KERNEL)))
		return NULL;
	memset(mousedev, 0, sizeof(struct mousedev));
	init_waitqueue_head(&mousedev->wait);

	mousedev->exist = 1;
	mousedev->minor = minor;
	mousedev_table[minor] = mousedev;

	mousedev->handle.dev = dev;
	mousedev->handle.handler = handler;
	mousedev->handle.private = mousedev;

	mousedev->devfs = input_register_minor("mouse%d", minor, MOUSEDEV_MINOR_BASE);

	if (mousedev_mix.open)
		input_open_device(&mousedev->handle);

//	printk(KERN_INFO "mouse%d: PS/2 mouse device for input%d\n", minor, dev->number);

	return &mousedev->handle;
}

static void mousedev_disconnect(struct input_handle *handle)
{
	struct mousedev *mousedev = handle->private;

	mousedev->exist = 0;

	if (mousedev->open) {
		input_close_device(handle);
	} else {
		if (mousedev_mix.open)
			input_close_device(handle);
		input_unregister_minor(mousedev->devfs);
		mousedev_table[mousedev->minor] = NULL;
		kfree(mousedev);
	}
}
	
static struct input_handler mousedev_handler = {
	event:		mousedev_event,
	connect:	mousedev_connect,
	disconnect:	mousedev_disconnect,
	fops:		&mousedev_fops,
	minor:		MOUSEDEV_MINOR_BASE,
};

static int __init mousedev_init(void)
{
	printk(KERN_INFO "mousedev.c: version " MOUSEDEV_VERSION "\n");
	input_register_handler(&mousedev_handler);

	memset(&mousedev_mix, 0, sizeof(struct mousedev));
	init_waitqueue_head(&mousedev_mix.wait);
	mousedev_table[MOUSEDEV_MIX] = &mousedev_mix;
	mousedev_mix.exist = 1;
	mousedev_mix.minor = MOUSEDEV_MIX;
	mousedev_mix.devfs = input_register_minor("mice", MOUSEDEV_MIX, MOUSEDEV_MINOR_BASE);

	printk(KERN_INFO "mice: PS/2 mouse device common for all mice\n");

	return 0;
}

static void __exit mousedev_exit(void)
{
	input_unregister_minor(mousedev_mix.devfs);
	input_unregister_handler(&mousedev_handler);
}

module_init(mousedev_init);
module_exit(mousedev_exit);

MODULE_AUTHOR("Vojtech Pavlik <vojtech@suse.cz>");
#ifndef __JEJ_DEBUG
MODULE_DESCRIPTION("Input driver to PS/2 or ImPS/2 device driver (LINUXWACOM-DEBUG)");
#else
MODULE_DESCRIPTION("Input driver to PS/2 or ImPS/2 device driver (LINUXWACOM)");
#endif
MODULE_LICENSE("GPL");

MODULE_PARM(xres, "i");
MODULE_PARM_DESC(xres, "Horizontal screen resolution");
MODULE_PARM(yres, "i");
MODULE_PARM_DESC(yres, "Vertical screen resolution");
